### PR TITLE
[RLlib] Rllib train pg

### PR DESCRIPTION
**Pull request by avnishn**
This PR is a prototype of how we would integrate our algorithms with ray train, and remove our own in house multi-gpu support. 

A few things I've noticed:
- it works in the multi agent case, although, I'm not sure how, since I thought that multiple DDP networks couldn't sit on the same process. CC @matthewdeng @amogkam 
- there seems to be a small amount of parameter drift after a few training iterations. I'm not sure if that was expected, so I need some guidance here. Is this expected? 
- Is it possible that we aren't applying the gradients in the exact same way across our different train workers? 
- It seems to provide stable learning, and there appears to be no problems with setting seed. =

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

I wrote an assertion that checks to see whether parameters are the same across the different training workers. It's what I used to catch the parameter drift. 

The parameter drift happens regardless of whether I'm doing multi-agent training (multiple ddp networks on one train worker) or not (only 1 ddp network on each train worker)

### CODE REVIEW

File: rllib/agents/pg/pg.py [hunk 1]
```diff
@@ -32,3 +86,228 @@ def get_default_config(cls) -> TrainerConfigDict:
...
+        if self._train_trainer:
+            return self._rllib_on_train_training_iteration_fn()
+        else:
+            return super(PGTrainer, self).training_iteration()
+
+    def _rllib_on_train_training_iteration_fn(self) -> ResultDict:
+        # Some shortcuts.
+        if not self._training_workers_started:
+            for idx, w in enumerate(self._training_workers):
+                w.start.remote(0, self._num_train_workers)
```
Comment (matthewdeng) [inline rllib/agents/pg/pg.py:101]: Should this use `idx` instead?
```suggestion
                w.start.remote(idx, self._num_train_workers)
```

File: rllib/agents/pg/pg.py [hunk 2]
```diff
@@ -32,3 +86,228 @@ def get_default_config(cls) -> TrainerConfigDict:
...
+            return
+        for pid in self._local_worker.policies_to_train:
+            model = self._local_worker.policy_map[pid].model
+            if isinstance(model, DDPTorchModelV2Wrapper) or not isinstance(
+                model, TorchModelV2
+            ):
+                continue
+            if torch.cuda.is_available():
+                model = DDPTorchModelV2Wrapper(
+                    model, device_ids=[rank], output_device=rank
```
Comment (matthewdeng) [inline rllib/agents/pg/pg.py:229]: Not sure if you're using (multi-node) GPU yet but this should be local rank (as opposed to global rank)

File: rllib/agents/pg/pg.py [hunk 3]
```diff
@@ -24,6 +50,34 @@ class PGTrainer(Trainer):
...
+            num_gpus_per_training_worker = config_copy["num_gpus_per_training_worker"]
+
+            self._train_trainer = train_trainer(
+                "torch",
+                logdir="/tmp/rllib_tmp",
+                num_workers=self._num_train_workers,
+                resources_per_worker={
+                    "CPU": num_cpus_per_training_worker,
+                    "GPU": num_gpus_per_training_worker,
+                },
```
Comment (matthewdeng) [inline rllib/agents/pg/pg.py:69]: I guess you're using CPU only now, but if you want to use GPU you'll need to add something like : `use_gpu=bool(num_gpus_per_training_worker)`
